core/templates/dev/head/exploration_editor/EditorServices.js

```diff
@@ -1952,3 +2017,310 @@ oppia.factory('explorationWarningsService', [
+                    .append(defaultOutcomeHtml));
+              } else if (defaultOutcomeChanges === 'edited') {
+                if (newValue.dest !== oldValue.dest) {
+                  defaultOutcomeHtml += (
+                    '<p>Destination: ' + newValue.dest + '</p>');
+                }
+                if (!angular.equals(newValue.feedback, oldValue.feedback)) {
+                  defaultOutcomeHtml += (
+                    '<p>Feedback: ' + newValue.feedback + '</p>');
+                }
+                stateWiseEditsMapping[stateName].push(
+                  angular.element('<div>Edited default outcome: </div>')
+                    .append(defaultOutcomeHtml));
+              } else if (defaultOutcomeChanges === 'deleted') {
+                stateWiseEditsMapping[stateName].push(
+                  angular.element('<div>Deleted default outcome</div>'));
+              }
+          };
+      }
+    });
+
+    for (var stateName in stateWiseEditsMapping) {
+      var stateChangesEl = angular.element(
+        '<li>Edits to state: ' + stateName + '</li>');
+      for (var stateEdit in stateWiseEditsMapping[stateName]) {
+        stateChangesEl.append(stateWiseEditsMapping[stateName][stateEdit]);
+      }
+      outerHtml.append(stateChangesEl);
+    }
+
+    return outerHtml;
+  };
+
+  return {
+    makeHumanReadable: makeHumanReadable
+  };
+}]);
+
+// Service for displaying different types of modals depending on the type of
+// response received as a result of the autosaving request.
+oppia.factory('autosaveInfoModalsService', [
+  '$modal', '$timeout', '$window', 'lostChangesService', 'explorationData',
+  function(
+    $modal, $timeout, $window, lostChangesService, explorationData) {
+    var _isNonStrictFailModalOpen = false;
+    var _isVersionMismatchModalOpen = false;
+
+    return {
+      showNonStrictValidationFailModal: function() {
+        $modal.open({
+          templateUrl: 'modals/saveValidationFail',
+          backdrop: true,
+          controller: [
+            '$scope', '$modalInstance',
+            function(
+              $scope, $modalInstance
+            ) {
+              $scope.close = function() {
+                $modalInstance.dismiss('cancel');
+              };
+            }
+          ]
+        }).result.then(function() {
+          _isNonStrictFailModalOpen = false;
+        }, function() {
+          _isNonStrictFailModalOpen = false;
+        });
+
+        _isNonStrictFailModalOpen = true;
+      },
+      isNonStrictFailModalOpen: function() {
+        return _isNonStrictFailModalOpen;
+      },
+      showVersionMismatchModal: function(lostChanges) {
+        $modal.open({
+          templateUrl: 'modals/saveVersionMismatch',
+          backdrop: true,
+          controller: [
+            '$scope', '$modalInstance',
+            function(
+              $scope, $modalInstance
+            ) {
+              $scope.cancel = function() {
```

Comment:
We don't want to give the option to cancel as we need to force changes to be discarded.